### ISSUE TITLE
release for scala 2.13.0-M5?

### ISSUE THREAD

**Issue by pjfanning** [akka/akka#25105]
https://www.scala-lang.org/download/2.13.0-M4.html - has been released

Relates to https://github.com/akka/akka/issues/24507 
Depends on ~https://github.com/scalatest/scalatest/issues/1367~: 3.0.6-SNAP1 should work

**Comment by raboof** [comment]
:+1: - also depends on https://github.com/scala/scala-java8-compat

**Comment by xuwei-k** [comment]
- [x] scala-java8-compat  https://oss.sonatype.org/content/repositories/releases/org/scala-lang/modules/scala-java8-compat_2.13.0-M4/
- [ ] ssl-config https://github.com/lightbend/ssl-config/pull/83
- [ ] scala-stm
- [ ] spray-json (only akka-docs project?)

**Comment by schmitch** [comment]
ssl-config has a 2.13.0-M4 release http://central.maven.org/maven2/com/typesafe/ssl-config-core_2.13.0-M4/0.2.4/.

**Comment by MasseGuillaume** [comment]
@raboof Is there anyone on the Akka team working on this? If not, I would like to do it. It would be a good occasion to battle test https://github.com/scala/scala-collection-compat and it's migration rules.

**Comment by xuwei-k** [comment]
> Is there anyone on the Akka team working on this?

https://github.com/akka/akka/pull/25267

**Comment by raboof** [comment]
@MasseGuillaume as @xuwei-k linked we've started exploring a bit, but your help would be greatly appreciated!

**Comment by SethTisue** [comment]
checking in, anything we should be discussing here...?

**Comment by ktoso** [comment]
I'll check out status and report back

**Comment by ktoso** [comment]
Okey so the PRs above are in progress and we need to continue the work here. 
I'm leaving this to @raboof to continue once he's back from vacation next week :)

**Comment by MasseGuillaume** [comment]
@ktoso I'm currently working on this. Not as fast as I could since we use this to validate our github.com/scala/scala-collection-compat project and increase the quality of our automatic migration rules. This is what I got out of this so far:

* Rewrite breakOut
* Rewrite col.to and col.to[CC] => col.to(CC)
* Rewrite Future.{onSuccess, OnFailure} => Future.onComplete
* Rewrite JavaConversion
* Rewrite from,to,until => range{From,To,Until}
* Rewrite domino foldRight and foldLeft patterns (`/:` and `:\`)
* Rewrite Traversable* to Iterable*

The only manual par remaining would be:

* Decide if a collection.Seq should be immutable or mutable. (Akka normally marks immutable.Seq)
* Custom collections (such as ByteString)

You can track my progress here: https://github.com/MasseGuillaume/akka/commits/2.13.0-M4

Unless you are blocked by this, I would say leave me at least 2 weeks, then I would definitively need your help for PR review, making unit tests pass.

**Comment by SethTisue** [comment]
I'm curious how close Guillaume's efforts got to something releasable. does the Akka team expect to be able to publish for M5 fairly soon after all your dependencies publish...?

**Comment by dwijnand** [comment]
Btw, I suggest https://github.com/SethTisue/akka/tree/community-build-2.13 as the branch to integrate.

**Comment by patriknw** [comment]
we can do another push next week

**Comment by raboof** [comment]
Yes, we are using this as well. It is indeed incredibly helpful, but as you mention there is still quite some extra work to be done, especially since we want to maintain binary incompatibility, and in some cases the migration tool will violate that.

**Comment by pjfanning** [comment]
@raboof I suspect that it won't be possible to be binary compatible and that you will need to do the changes on an akka 2.6 code line.

**Comment by raboof** [comment]
@pjfanning that would be quite disappointing. Can you give an example of a construct that you suspect cannot be handled in a binary compatible way?

**Comment by SethTisue** [comment]
dozens of libraries have now published for M4 and/or M5 and I don't think any have yet needed to break bincompat to do it. perhaps there are counterexamples, but it isn't the norm.

Akka is big and could well be an exception, but it's worth being optimistic about.

I wonder MiMa still passes on 2.11.x/2.12.x when @MasseGuillaume's patches are applied.

**Comment by raboof** [comment]
(I plan to look into this more later this week)

**Comment by pjfanning** [comment]
https://github.com/slick/slick/pull/1966/files - this is an example from applying the migration to Slick code base. Note that `Traversable` is replaced with `Iterable`, `TraversableOnce` is replaced with `IterableOnce` and that `CanBuildFrom` is replaced by `Factory`. It also introduces a dependency on the scala-collection-compat library.

**Comment by SethTisue** [comment]
thanks for the list. the libraries I've seen pull this off have used other workarounds rather than make the changes you list. they typically involve using small amounts of Scala-version-specific source code.

**Comment by raboof** [comment]
@pjfanning since `Traversable` is still available on 2.13 (albeit deprecated), perhaps we could just switch those changes back when they happen in the API. If that is not possible, perhaps we can use the technique from https://github.com/scala/scala-collection-compat/blob/master/compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala#L151?

I'm not sure `CanBuildFrom`/`Factory` is used in the API in many cases, but the fact that on 2.11/2.12 `Factory` actually is a `CanBuildFrom` on 2.11 might help (https://github.com/scala/scala-collection-compat/blob/master/compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala#L19)

Adding the dependency on scala-collection-compat should not be a problem (as soon as the time is right to start promising binary compatibility for that library itself, of course)

**Comment by SethTisue** [comment]
> Adding the dependency on scala-collection-compat should not be a problem

you might consider avoiding it if possible, though, since as we all know, mo dependencies mo problems, especially downstream. most library maintainers so far are choosing not to use it.

**Comment by raboof** [comment]
@SethTisue that's interesting - I would love to avoid it but got the impression that a big reason for the library to exist in the first place would be to allow cross-building code that uses `CanBuildFrom` on 2.12. Of course an alternative might be to duplicate such code and provide different implementations for 2.12 and 2.13, but that seems like a ton of work that the library was designed to alleviate.

**Comment by jrudolph** [comment]
Could it make sense for Akka to copy only parts from the compat library somewhere into our codebase (if it's mostly about avoiding dependencies)?

**Comment by SethTisue** [comment]
>  Of course an alternative might be to duplicate such code and provide different implementations for 2.12 and 2.13, but that seems like a ton of work

for at least some particular libraries it hasn't been that bad, see e.g.:

* https://github.com/scala/scala-xml/blob/master/shared/src/main/scala-2.11-2.12/scala/xml/ScalaVersionSpecific.scala
* https://github.com/scala/scala-xml/blob/master/shared/src/main/scala-2.13/scala/xml/ScalaVersionSpecific.scala

> Could it make sense for Akka to copy only parts from the compat library somewhere into our codebase (if it's mostly about avoiding dependencies)?

perhaps, yes. copy and paste is better if the code involved isn't too long or involved.

**Comment by SethTisue** [comment]
@szeiger and @julienrf may have perspectives on this

**Comment by dwijnand** [comment]
I think while:

* #26043 is still unmerged (the publishings were off of that, also it needs a rebase), and
* not all modules of Akka were published afaik

maybe we should, calling this "effectively done" and create follow-up tickets for the follow-up work, that interest parties can subscribe to if interested.  For instance, I'd be interested in tracking when 2.13 support is entirely merged into the master build/CI and by default a part of Akka releases, rather than in opt-in/"someone do it" state.
